src/test/SampleCustomLoader/sampleCustLoader.cpp

```diff
@@ -0,0 +1,320 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include <chrono>
+#include <fstream>
+#include <future>
+#include <iostream>
+#include <map>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <assert.h>
+#include <rapidjson/document.h>
+
+#include "../../customloaderinterface.hpp"
+
+using namespace std;
+using namespace rapidjson;
+using namespace ovms;
+
+#define PATH_SIZE 10
+#define RSIZE_MAX_STR 4096
+
+#define CUSTLOADER_OK 1
+#define CUSTLOADER_ERROR 40
+
+typedef pair<string, int> model_id_t;
+
+class custSampleLoader : public CustomLoaderInterface {
+private:
+    vector<model_id_t> models_blacklist;
+    vector<model_id_t> models_loaded;
+
+protected:
+    int extract_input_params(char* basePath, int version, char* loaderOptions, char** binFile,
+        char** xmlFile);
+    int load_files(char* binFile, char* xmlFile, char** xmlBuffer, char** binBuffer, int* xmlLen,
+        int* binLen);
+
+    std::promise<void> exitSignal;
+    std::future<void> futureObj;
+    int watchIntervalSec = 0;
+    bool watcherStarted = false;
+    std::thread watcher_thread;
+
+public:
+    custSampleLoader();
+    ~custSampleLoader();
+    int loaderInit(char* loader_path);
+    void loaderDeInit();
+    int unloadModel(const char* modelName, int version);
+    int loadModel(const char* modelName, const char* basePath, const int version,
+        const char* loaderOptions, char** xmlBuffer, int* xmlLen, char** binBuffer,
+        int* binLen);
+    bool getModelBlacklistStatus(const char* modelName, int version);
+
+    void threadFunction(std::future<void> futureObj);
+    void startWatcher(int intervalSec);
+    void watcherJoin();
+};
+
+extern "C" CustomLoaderInterface* create() {
+    return new custSampleLoader();
+}
+
+extern "C" void destroy(CustomLoaderInterface* p) {
+    delete p;
+}
+
+custSampleLoader::custSampleLoader() {
+    cout << "custSampleLoader: Instance of Custom SampleLoader created" << endl;
+}
+
+custSampleLoader::~custSampleLoader() {
+    cout << "custSampleLoader: Instance of Custom SampleLoader deleted" << endl;
+}
+
+int custSampleLoader::loaderInit(char* loader_path) {
+    cout << "custSampleLoader: Custom loaderInit" << loader_path << endl;
+    // if error return CUSTLOADER_INIT_FAIL;
+    return CUSTLOADER_OK;
+}
+
+int custSampleLoader::load_files(char* binFile, char* xmlFile, char** xmlBuffer,
+    char** binBuffer, int* xmlLen, int* binLen) {
+    streampos size;
+    ifstream bfile(binFile, ios::in | ios::binary | ios::ate);
+    if (bfile.is_open()) {
+        size = bfile.tellg();
+        *binLen = size;
+        *binBuffer = new char[size];

```

Comment:
Please explain lifetime of a binBuffer, who is responsible of releasing it and if this way (delete []) is properly implemented and explained. This may also assume compatible stdlibc++ version across shared objects, which may or may not be the case.

Reply:
Delete has been implemented in the OVMS code